chemprop/v2/cli/utils/parsing.py

```diff
@@ -0,0 +1,215 @@
+
+    if target_cols is None:
+        target_cols = list(set(df.columns) - set(input_cols) - set(ignore_cols or []))
+
+    Y = df[target_cols]
+    weights = None if weight_col is None else df[weight_col].to_numpy()
+
+    if bounded:
+        lt_mask = Y.applymap(lambda x: "<" in x).to_numpy()
+        gt_mask = Y.applymap(lambda x: ">" in x).to_numpy()
+        Y = Y.applymap(lambda x: x.strip("<").strip(">")).astype(float).to_numpy()
+    else:
+        Y = Y.to_numpy()
+        lt_mask = None
+        gt_mask = None
+
+    return smiss, rxnss, Y, weights, lt_mask, gt_mask
+
+
+def make_datapoints(
+    smiss: list[list[str]] | None,
+    rxnss: list[list[str]] | None,
+    Y: np.ndarray,
+    weights: np.ndarray | None,
+    lt_mask: np.ndarray | None,
+    gt_mask: np.ndarray | None,
+    X_f: np.ndarray | None,
+    V_fs: list[np.ndarray] | None,
+    E_fs: list[np.ndarray] | None,
+    V_ds: list[np.ndarray] | None,
+    features_generators: str | None,
+    keep_h: bool,
+    add_h: bool,
+) -> tuple[list[list[MoleculeDatapoint]], list[list[ReactionDatapoint]]]:
+    """Make the :class:`MoleculeDatapoint`s and :class:`ReactionDatapoint`s for a given
+    dataset.
+    Parameters
+    ----------
+
+    Returns
+    -------
+    list[list[MoleculeDatapoint]]
+        a list of lists of :class:`MoleculeDatapoint`s of shape ``j x n``, where ``j`` is the
+        number of molecule components per datapoint and ``n`` is the total number of datapoints
+    list[list[ReactionDatapoint]]
+        a list of lists of :class:`MoleculeDatapoint`s of shape ``k x n``, where ``k`` is the
+        number of reaction components per datapoint and ``n`` is the total number of datapoints
+    .. note::
+        either ``j`` or ``k`` may be 0, in which case the corresponding list will be empty.
+    """
+    if smiss is None and rxnss is None:
+        raise ValueError("args 'smiss' and 'rnxss' were both `None`!")
+    elif rxnss is None:
+        N = len(smiss)
+        rxnss = []
+    elif smiss is None:
+        N = len(rxnss)
+        smiss = []
+    elif len(smiss) != len(rxnss):
+        raise ValueError(
+            f"args 'smiss' and 'rxnss' must have same length! got {len(smiss)} and {len(rxnss)}"
+        )
+
+    weights = np.ones(N) if weights is None else weights
+    gt_mask = [None] * N if gt_mask is None else gt_mask
+    lt_mask = [None] * N if lt_mask is None else lt_mask
+    X_f = [None] * N if X_f is None else X_f
+    V_fs = [None] * N if V_fs is None else V_fs
+    E_fs = [None] * N if E_fs is None else E_fs
+    V_ds = [None] * N if V_ds is None else V_ds
+
+    mol_data = [
+        [
+            MoleculeDatapoint.from_smi(
+                smis[i],
+                keep_h=keep_h,
+                add_h=add_h,
+                y=Y[i],
+                weight=weights[i],
+                gt_mask=gt_mask[i],
+                lt_mask=lt_mask[i],
+                x_f=X_f[i],
+                mfs=features_generators,
```

Comment:
I think this will work correctly when `features_generators` is `None`, but it couldn't the case where a user passes in a feature generator for mol and a feature generator for rxn. Maybe separate keyword variables are needed for specifying feature generator for mol and for rxn?